erpnext/erpnext_integrations/doctype/plaid_settings/plaid_connector.py

```diff
@@ -2,78 +2,87 @@
-			environment=self.config.get("plaid_env")
-			)
-
 		self.access_token = access_token
+		self.settings = frappe.get_single("Plaid Settings")
+		self.products = ["auth", "transactions"]
+		self.client_name = frappe.local.site
+		self.client = plaid.Client(
+			client_id=self.settings.plaid_client_id,
+			secret=self.settings.get_password('plaid_secret'),
+			environment=self.settings.plaid_env,
+			api_version='2019-05-29'
+		)
 
 	def get_access_token(self, public_token):
 		if public_token is None:
 			frappe.log_error(_("Public token is missing for this bank"), _("Plaid public token error"))
-
 		response = self.client.Item.public_token.exchange(public_token)
 		access_token = response['access_token']
-
 		return access_token
 
+	def get_link_token(self):
+		token_request = {
+			"client_name": self.client_name,
+			"client_id": self.settings.plaid_client_id,
+			"secret": self.settings.plaid_secret,
+			"products": self.products,
+			# only allow Plaid-supported languages (LAST: Sep-12-2020)
+			"language": frappe.local.lang if frappe.local.lang in ["en", "fr", "es", "nl"] else "en",
+			"country_codes": ["US"],
+			"user": {
+				"client_user_id": frappe.generate_hash(frappe.session.user, length=32)
+			}
+		}
+
+		try:
+			response = self.client.LinkToken.create(token_request)
+		except InvalidRequestError as e:
+			frappe.log_error(frappe.get_traceback(), _("Plaid invalid request error"))
+			frappe.msgprint(_("Please check your Plaid client ID and secret values"))
+		except APIError as e:
+			frappe.log_error(frappe.get_traceback(), _("Plaid authentication error"))
+			frappe.msgprint({"title": _("Authentication Failed"), "message": e, "raise_exception": 1, "indicator": 'red'})
+		else:
+			return response["link_token"]
+
 	def auth(self):
 		try:
 			self.client.Auth.get(self.access_token)
 			print("Authentication successful.....")
 		except ItemError as e:
 			if e.code == 'ITEM_LOGIN_REQUIRED':
 				pass
-			else:
-				pass
 		except APIError as e:
 			if e.code == 'PLANNED_MAINTENANCE':
 				pass
-			else:
-				pass
 		except requests.Timeout:
 			pass
 		except Exception as e:
 			print(e)
 			frappe.log_error(frappe.get_traceback(), _("Plaid authentication error"))
-			frappe.msgprint({"title": _("Authentication Failed"), "message":e, "raise_exception":1, "indicator":'red'})
+			frappe.msgprint({"title": _("Authentication Failed"), "message": e, "raise_exception": 1, "indicator": 'red'})
 
 	def get_transactions(self, start_date, end_date, account_id=None):
 		try:
 			self.auth()
 			if account_id:
 				account_ids = [account_id]
-
 				response = self.client.Transactions.get(self.access_token, start_date=start_date, end_date=end_date, account_ids=account_ids)
-
 			else:
 				response = self.client.Transactions.get(self.access_token, start_date=start_date, end_date=end_date)
 
 			transactions = response['transactions']
-
 			while len(transactions) < response['total_transactions']:
```

Comment:
Oh wow I haven't seen a while used in a..._**while**_. Would this be better off just being a loop through `response['total_transactions']`?

Reply:
They seemed to have picked that up from Plaid's official docs, I think even if we do a loop, there'll have to be an `if` length check anyway later on.

I'll compare the responses and see if it can be replaced though.

Reply:
Tested this out, `while` seems to be pretty decent for parsing the transactions here, gonna keep it as is